From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Fri, 11 Jul 2025 12:29:27 +0800
Subject: [PATCH] Use fernflower's result for getDismountLocationInDirection in
 AbstractHorse

A weird but working way to fix the error of vineflower

diff --git a/net/minecraft/world/entity/animal/horse/AbstractHorse.java b/net/minecraft/world/entity/animal/horse/AbstractHorse.java
index 31ed5d7cfbda1a1f98ebfc89f5cbc65332cabdae..753b8a40933875da2d9517e27302c86abc263724 100644
--- a/net/minecraft/world/entity/animal/horse/AbstractHorse.java
+++ b/net/minecraft/world/entity/animal/horse/AbstractHorse.java
@@ -979,33 +979,38 @@ public abstract class AbstractHorse extends Animal implements HasCustomInventory
     }
 
     @Nullable
-    private Vec3 getDismountLocationInDirection(Vec3 direction, LivingEntity passenger) {
-        double d = this.getX() + direction.x;
-        double d1 = this.getBoundingBox().minY;
-        double d2 = this.getZ() + direction.z;
+    private Vec3 getDismountLocationInDirection(Vec3 vec3, LivingEntity livingEntity) {
+        double d = this.getX() + vec3.x;
+        double e = this.getBoundingBox().minY;
+        double f = this.getZ() + vec3.z;
         BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
+        com.google.common.collect.UnmodifiableIterator<Pose> var10 = livingEntity.getDismountPoses().iterator();
 
-        for (Pose pose : passenger.getDismountPoses()) {
-            mutableBlockPos.set(d, d1, d2);
-            double d3 = this.getBoundingBox().maxY + 0.75;
+        while(var10.hasNext()) {
+            Pose pose = (Pose)var10.next();
+            mutableBlockPos.set(d, e, f);
+            double g = this.getBoundingBox().maxY + (double)0.75F;
 
-            do {
-                double blockFloorHeight = this.level().getBlockFloorHeight(mutableBlockPos);
-                if (mutableBlockPos.getY() + blockFloorHeight > d3) {
+            while(true) {
+                double h = this.level().getBlockFloorHeight(mutableBlockPos);
+                if ((double)mutableBlockPos.getY() + h > g) {
                     break;
                 }
 
-                if (DismountHelper.isBlockFloorValid(blockFloorHeight)) {
-                    AABB localBoundsForPose = passenger.getLocalBoundsForPose(pose);
-                    Vec3 vec3 = new Vec3(d, mutableBlockPos.getY() + blockFloorHeight, d2);
-                    if (DismountHelper.canDismountTo(this.level(), passenger, localBoundsForPose.move(vec3))) {
-                        passenger.setPose(pose);
-                        return vec3;
+                if (DismountHelper.isBlockFloorValid(h)) {
+                    AABB aABB = livingEntity.getLocalBoundsForPose(pose);
+                    Vec3 vec32 = new Vec3(d, (double)mutableBlockPos.getY() + h, f);
+                    if (DismountHelper.canDismountTo(this.level(), livingEntity, aABB.move(vec32))) {
+                        livingEntity.setPose(pose);
+                        return vec32;
                     }
                 }
 
                 mutableBlockPos.move(Direction.UP);
-            } while (!(mutableBlockPos.getY() < d3));
+                if (!((double)mutableBlockPos.getY() < g)) {
+                    break;
+                }
+            }
         }
 
         return null;
